server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -340,15 +360,59 @@ private void loadOrExecuteQueryPhase(final ShardSearchRequest request, final Sea
     }
 
     public void executeQueryPhase(ShardSearchRequest request, SearchShardTask task, ActionListener<SearchPhaseResult> listener) {
-        rewriteShardRequest(request, ActionListener.map(listener, r -> executeQueryPhase(r, task)));
+        assert request.isMatchNoDocsReturnNullResponse() == false || request.numberOfShards() > 1
+            : "empty responses require more than one shard";
+        rewriteShardRequest(request, ActionListener.wrap(
+            context -> {
+                try {
+                    ShardSearchRequest rewritten = context.request;
+                    if (rewritten.isMatchNoDocsReturnNullResponse()
+                            && canRewriteToMatchNone(rewritten.source())
+                            && rewritten.source().query() instanceof MatchNoneQueryBuilder) {
+                        onMatchNoDocs(context, listener);
+                    } else {
+                        // fork the execution in the search thread pool and wraps the searcher
+                        // to execute the query
+                        context.wrapSearcher().execute(() -> {
+                            final SearchPhaseResult result;
+                            try {
+                                result = executeQueryPhase(context, task);
+                            } catch (Exception exc) {
+                                listener.onFailure(exc);
+                                return;
+                            }
+                            listener.onResponse(result);
+                        });
+                    }
+                } catch (Exception exc) {
+                    // if the execution is rejected we need to close the searcher
+                    IOUtils.closeWhileHandlingException(context.searcher);
+                    listener.onFailure(exc);
+                }
+            }, listener::onFailure));
+    }
+
+    private void onMatchNoDocs(SearchRewriteContext rewriteContext, ActionListener<SearchPhaseResult> listener) {
+        // creates a lightweight search context that we use to inform context listeners
+        // before closing
+        SearchContext searchContext = createSearchContext(rewriteContext, defaultSearchTimeout);
+        try (searchContext) {
+            onNewContext(searchContext);
```

Comment:
Can we do skip these steps entirely? If so, we might not need to acquire an engine searcher, which potentially triggers a refresh.

Reply:
I added this to ensure that frozen indices are "unregistered" (nullify) when the context is freed. It should be easier when we move to a simple reader context ;). 
For the refresh, I think this is required since we have to handle the search-idle case that automatically refreshes on the first query.

Reply:
Ok, let's make this change later with the reader context.